### PR TITLE
feat(tariscript): adds ToRistrettoPoint op-code

### PR DESCRIPTION
**Pull request by sdbondi**
Description
---
- Adds ToRistrettoPoint opcodes
- Added scalar stack item

Motivation and Context
---
As per RFC-202
Ref https://github.com/tari-project/rfcs/pull/15
Ref https://github.com/tari-project/tari/pull/4742

How Has This Been Tested?
---
Additional unit tests + tests updated

### CODE REVIEW

File: infrastructure/tari_script/src/op_codes.rs [hunk 1]
```diff
@@ -116,6 +116,7 @@ pub const OP_CHECK_MULTI_SIG_VERIFY: u8 = 0xaf;
 pub const OP_HASH_BLAKE256: u8 = 0xb0;
 pub const OP_HASH_SHA256: u8 = 0xb1;
 pub const OP_HASH_SHA3: u8 = 0xb2;
+pub const OP_TO_RISTRETTO_POINT: u8 = 0xb3;
```
Comment (delta1) [inline infrastructure/tari_script/src/op_codes.rs:120]: https://github.com/tari-project/tari/pull/4742/files#diff-307c63687b44168345b04870922da3ee590273f9c3cc8b57795d1992d3706386R119

Comment (sdbondi) [reply infrastructure/tari_script/src/op_codes.rs:120]: Thanks :eyes:

**Review by hansieodendaal — DISMISSED** [review]
Nice, just noted today this is missing.

LGTM

File: infrastructure/tari_script/src/script.rs [hunk 2]
```diff
@@ -537,6 +539,19 @@ impl TariScript {
...
+
+    fn handle_to_ristretto_point(&self, stack: &mut ExecutionStack) -> Result<(), ScriptError> {
+        let item = stack.pop().ok_or(ScriptError::StackUnderflow)?;
+        let scalar = match &item {
+            StackItem::Hash(hash) => hash.as_slice(),
+            StackItem::Scalar(scalar) => scalar.as_slice(),
+            _ => return Err(ScriptError::IncompatibleTypes),
+        };
+        let scalar = RistrettoSecretKey::from_bytes(scalar).map_err(|_| ScriptError::InvalidData)?;
+        let ristretto_pk = RistrettoPublicKey::from_secret_key(&scalar);
```
Comment (willyrgf) [inline infrastructure/tari_script/src/script.rs]: May improve readability.

```suggestion
        let ristretto_sk = RistrettoSecretKey::from_bytes(scalar).map_err(|_| ScriptError::InvalidData)?;
        let ristretto_pk = RistrettoPublicKey::from_secret_key(&ristretto_sk);
```

File: infrastructure/tari_script/src/stack.rs [hunk 3]
```diff
@@ -52,11 +52,13 @@ pub const TYPE_HASH: u8 = 2;
 pub const TYPE_COMMITMENT: u8 = 3;
 pub const TYPE_PUBKEY: u8 = 4;
 pub const TYPE_SIG: u8 = 5;
+pub const TYPE_SCALAR: u8 = 6;
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum StackItem {
     Number(i64),
     Hash(HashValue),
+    Scalar([u8; 32]),
```
Comment (willyrgf) [inline infrastructure/tari_script/src/stack.rs]: Should we have something like `HashValue` to scalar? Like a `pub type ScalarValue = [u8; 32];` in `infrastructure/tari_script/src/op_codes.rs`.

**Comment by willyrgf** [comment]
What do you think about this kind of approach to `StackItem` values and similar cases? https://github.com/willyrgf/tari/commit/8e67de1e178f1dd465aee5f1785c9f0afc38de14

File: infrastructure/tari_script/src/stack.rs (hunk 3 above)
Comment (sdbondi) [reply infrastructure/tari_script/src/stack.rs]: Yeah agree with that, will change

**Comment by sdbondi** [comment]
> What do you think about this kind of approach to `StackItem` values and similar cases? [willyrgf@8e67de1](https://github.com/willyrgf/tari/commit/8e67de1e178f1dd465aee5f1785c9f0afc38de14)

Yeah that is a little nicer, however I'd favour the current approach over introducing another dependency.

File: infrastructure/tari_script/src/script.rs (hunk 2 above)
Comment (sdbondi) [reply infrastructure/tari_script/src/script.rs]: Will change :ok_hand: - the RistrettoSecret key part is more of an intermediate step so not technically a secret

**Comment by willyrgf** [comment]
> > What do you think about this kind of approach to `StackItem` values and similar cases? [willyrgf@8e67de1](https://github.com/willyrgf/tari/commit/8e67de1e178f1dd465aee5f1785c9f0afc38de14)
> 
> 
> 
> Yeah that is a little nicer, however I'd favour the current approach over introducing another dependency.

Yes, it's make sense. Maybe in the future with [`core::mem::variant_count`](https://github.com/rust-lang/rust/issues/73662).

**Review by stringhandler — APPROVED** [review]

**Comment by stringhandler** [comment]
In future we will need to hard fork these in, but since this will not be used immediately, there should be enough time for nodes to upgrade

**Pull request merged**